Add snapshot encoding for slices

dario23 opened the pull request.

dario23 committed:
Implement slices in pure functions and specifications

dario23 committed:
Add testcase for slices in pure functions and specifications

Aurel300 commented on prusti-viper/src/encoder/pure_function_encoder.rs:
```diff
@@ -1345,9 +1361,17 @@ impl<'p, 'v: 'p, 'tcx: 'v> BackwardMirInterpreter<'tcx>
                             _ => {
                                 return Err(SpannedEncodingError::unsupported(
-                                    "checking the length of anything but arrays in pure code is not implemented yet",
+                                    "checking the length of anything but arrays and slices in pure code is not implemented yet",
```
I guess based on [`Rvalue::Len` documentation](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.Rvalue.html#variant.Len), the variant should never be emitted for anything other than arrays or slices?

dario23 replied:
hm, true. so go with `SpannedEncodingError::internal` then? or full-on `unreachable!()`?

Aurel300 replied:
`unreachable!` should be fine, with an appropriate comment.

dario23 committed:
Register bounds check ErrorCtxt for slices as well

dario23 committed:
Add pure function slice out-of-bounds testcase

dario23 committed:
Length request never happens on anything other than arrays or slices

dario23 committed:
Also trigger on lookup call for slice read axiom

dario23 committed:
Rework conditions, len encoding

dario23 committed:
Encode snapshotting operation for unsize-cast in pure code

dario23 committed:
Add two testcases for slices in pure code

Aurel300 commented on prusti-tests/tests/verify/pass/arrays/slice-sorted.rs:
```diff
@@ -0,0 +1,24 @@
+    }
+}
+
+/*
+#[requires(sorted(&a[0..2]))]
+fn requires_sorted_start(a: [i32; 3]) {
+    assert!(a[0] <= a[1]);
+}
+*/
```
Is this meant to be commented out?

dario23 replied:
yes; the `&a[0..2]` part doesn't work yet (only the unsize-cast that turns a full array into a slice like happens for `arr.len()`).

now that i think about it, i guess implementing full slicing (at least of arrays) makes sense within this PR.

dario23 replied:
alright, within snapshots we can now handle slicing of both arrays and slices :-)

dario23 committed:
Add slice split_at testcase

This wraps the built-in slice split_at method, to show we have all the
tools needed to ad specifications for that now.

We can't properly use #[extern_spec], because that doesn't work with
`impl [T]` for slices yet.

Aurel300 commented on prusti-tests/tests/verify/pass/arrays/slice-split.rs:
```diff
@@ -0,0 +1,19 @@
+    let (s0, s1) = split_at(&a, 7);
+
+    assert!(s0.len() == 7);
+    assert!(s1.len() == 6);
```
Given the postconditions, you should also be able to prove assertions about concrete values (e.g. `assert!(s0[2] == 0);`), is that right? If so, please add an assertion like that.

dario23 committed:
Add assertion about splitted slice contents

dario23 committed:
use vir::Function::apply

shorter and less error-prone

dario23 committed:
implement array snapshot slicing

dario23 committed:
Add slice_helper for slices as well

dario23 committed:
Add another testcase

Aurel300 commented on prusti-viper/src/encoder/pure_function_encoder.rs:
```diff
@@ -1345,15 +1406,62 @@ impl<'p, 'v: 'p, 'tcx: 'v> BackwardMirInterpreter<'tcx>
+
+                                state.substitute_value(&opt_lhs_value_place.unwrap(), snap_len);
                             }
+                            _ => span_bug!(span, "length should only be requested on arrays or slices"),
```
Does a `span_bug!` turn into a panic? If so, can we use an `internal` error instead?

dario23 replied:
i think it does, just like `unreachable!()` which we thought about earlier. what made me think more towards `span_bug!` instead of an internal error is that as far as we understand it, if we encounter anything else than array or slice here, it's a bug in `rustc` more than in prusti.

Aurel300 replied:
Ah, you're right, it's the `Len` thing. Never mind then, `span_bug` to panic makes sense.
